include/CObjectCache.h

```diff
@@ -70,191 +122,442 @@ namespace impl
 
 		inline size_t getSize() const { return m_container.size(); }
     };
-}
 
-template<
-    typename K,
-    typename T,
-    template<typename...> class ContainerT_T = std::vector,
-    bool = impl::is_same_templ<ContainerT_T, std::vector>::value
->
-class CObjectCache;
+    template<template<typename...> class ContainerT_T, typename ContainerT, bool ForMultiCache, bool IsAssocContainer = impl::is_assoc_container<ContainerT_T>::value>
+    struct CPreInsertionVerifier;
+    template<template<typename...> class ContainerT_T, typename ContainerT, bool IsAssocContainer>
+    struct CPreInsertionVerifier<ContainerT_T, ContainerT, true, IsAssocContainer>
+    {
+        template<typename ...Ts>
+        static bool verify(Ts...) { return true; }
+    };
+    template<template<typename...> class ContainerT_T, typename ContainerT>
+    struct CPreInsertionVerifier<ContainerT_T, ContainerT, false, false>
+    {
+        static bool verify(const ContainerT& _container, const typename ContainerT::iterator& _itr, const typename ContainerT::value_type::first_type& _key)
+        {
+            if (_itr != std::cend(_container) && !(_key < _itr->first)) // used `<` instead of `==` operator here to keep consistency with std::map (so key type doesn't need to define operator==)
+                return false;
 
-template<
-    typename K,
-    typename T,
-    template<typename...> class ContainerT_T
->
-class CObjectCache<K, T, ContainerT_T, true> : public impl::CObjectCacheBase<ContainerT_T, std::pair<K, T*>>
-{
-    using ValueType = std::pair<K, T*>;
+            return true;
+        }
+    };
+    template<template<typename...> class ContainerT_T, typename ContainerT>
+    struct CPreInsertionVerifier<ContainerT_T, ContainerT, false, true>
+    {
+        static bool verify(const std::pair<typename ContainerT::iterator, bool>& _insertionRes)
+        {
+            return _insertionRes.second;
```

Comment:
I don't understand this, why is a null `second` invalid?

Reference:
https://github.com/buildaworldnet/IrrlichtBAW/issues/123#issuecomment-409851961
https://github.com/buildaworldnet/IrrlichtBAW/commit/ed1029bd1995e07686dbd57c212d3a1ae8281fac

Reply:
`second` is `bool` here. It's the pair returned by non-multi assoc container from `std`